duckchat/duckchat-impl/src/main/java/com/duckduckgo/duckchat/impl/RealDuckChat.kt

```diff
@@ -0,0 +1,110 @@
+import com.squareup.moshi.JsonAdapter
+import com.squareup.moshi.Moshi
+import javax.inject.Inject
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.withContext
+
+interface DuckChatInternal : DuckChat {
+    /**
+     * Stores setting to determine whether the DuckChat should be shown in browser menu.
+     * Sets IO dispatcher.
+     */
+    suspend fun setShowInBrowserMenu(showDuckChat: Boolean)
+
+    /**
+     * Observes whether DuckChat should be shown in browser menu based on user settings and remote config flag
+     */
+    fun observeShowInBrowserMenu(): Flow<Boolean>
+}
+
+data class DuckChatSettingJson(
+    val aiChatURL: String,
+)
+
+@ContributesBinding(AppScope::class, boundType = DuckChat::class)
+@ContributesBinding(AppScope::class, boundType = DuckChatInternal::class)
+class RealDuckChat @Inject constructor(
+    private val duckChatFeatureRepository: DuckChatFeatureRepository,
+    private val duckChatFeature: DuckChatFeature,
+    private val moshi: Moshi,
+    private val dispatchers: DispatcherProvider,
+    private val globalActivityStarter: GlobalActivityStarter,
+    private val context: Context,
+) : DuckChatInternal {
+
+    private val jsonAdapter: JsonAdapter<DuckChatSettingJson> by lazy {
+        moshi.adapter(DuckChatSettingJson::class.java)
+    }
+
+    override suspend fun isEnabled(): Boolean = withContext(dispatchers.io()) {
+        duckChatFeature.self().isEnabled()
+    }
+
+    override suspend fun setShowInBrowserMenu(showDuckChat: Boolean) = withContext(dispatchers.io()) {
+        duckChatFeatureRepository.setShowInBrowserMenu(showDuckChat)
+    }
+
+    override fun observeShowInBrowserMenu(): Flow<Boolean> {
+        return duckChatFeatureRepository.observeShowInBrowserMenu().map {
+            it && duckChatFeature.self().isEnabled()
+        }
+    }
+
+    override suspend fun showInBrowserMenu(): Boolean = withContext(dispatchers.io()) {
+        duckChatFeatureRepository.shouldShowInBrowserMenu() && duckChatFeature.self().isEnabled()
+    }
+
+    override fun openDuckChat() {
+        val link = duckChatFeature.self().getSettings()?.let {
+            runCatching {
+                val settingsJson = jsonAdapter.fromJson(it)
+                settingsJson?.aiChatURL
+            }.getOrDefault(DUCK_CHAT_WEB_LINK)
+        } ?: DUCK_CHAT_WEB_LINK
+
+        val intent = globalActivityStarter.startIntent(
+            context,
+            WebViewActivityWithParams(
+                url = link,
+                screenTitle = context.getString(R.string.duck_chat_title),
+            ),
+        )
+
+        intent?.let {
+            it.flags = Intent.FLAG_ACTIVITY_NEW_TASK
+            context.startActivity(it)
+        }
+    }
+
+    companion object {
+        /** Default link to DuckChat that identifies Android as the source */
+        private const val DUCK_CHAT_WEB_LINK = "https://duckduckgo.com/?q=DuckDuckGo+AI+Chat&ia=chat&duckai=5"
```

Comment:
Should we add this to the config or will it never change?

Reply:
It’s in the config already :) This is just the default in case something is wrong with the config.

Reply:
Ah, that’s the default one. Makes sense 😁